### ISSUE TITLE
new rule proposal: no-falsy-string-init

### ISSUE THREAD

**Issue by dgreene1** [typescript-eslint/typescript-eslint#3230]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.
^^^ and yes, I see that I can almost use 'no-restricted-syntax' to accomplish this. See my workaround section to see why I think a truly new rule might be warranted.

# Proposal

A new rule called `'no-falsy-string-init'` (or whatever the community thinks would be a better name) that would help developers to declare a variable with a type that clarifies the variables optional nature.

## What this rule would prevent

```ts
let myStr: string = ''; // error  set the string to string | undefined     no-falsy-string-init
```

## What this rule would allow

```ts
let myString: string | undefined = '';
let myString: string | undefined;
```

# Real life scenario of the context

Think about situations like:
```ts
function getUserById(id: string){
  return http.get(`/user/${id}`);
}

const myId: string = '';

const aUser = getUserById(myId) // <-- Typescript can't help the dev since it thinks myId is truthy
```
That code is going to cause the API to return a 404 error or maybe even a BAD REQUEST error since there is no user with id ''But if we had a linter rule preventing strings being set to '' then Typescript itself would say "Hey buddy, you can't pass myId to getUserById until you've checked that it's truthy."

But after the linter rule (that I'm currently proposing) helps the dev to clarify their type, then Typescript can help the developer to get the rest of the way there.

i.e. after you fixed the linter error, you would see this:
```ts
function getUserById(id: string){
  return http.get(`/user/${id}`);
}

const myId: string | undefined = ''; // this is allowed now

const aUser = getUserById(myId) // Error myId might be undefined // Yay!
```

# Current workarounds

I know that I can use `'no-restricted-syntax'` to create a rule that gets close to this but:
1. I think the community might benefit from an easy to use rule as opposed to having to remember the AST syntax
2. I couldn't really figure out how to restrict this rule from firing on only `VariableDeclarator`s that have required `string` for the `typeAnnotation`. I got close, but it would be much easier to write this and test it within the testing framework that this repository has included.

But for those who want the cheap (but noisy route), here's a linter rule that will error on all empty string values. It's not a great workaround since "view code" has some valid cases for empty strings even though the data layer of an application rarely has a good reason for using empty string.
```json
'no-restricted-syntax': [
      'error',
      {
        selector: "Literal[value='']",
        message:
          'Please set the variable to undefined instead. If that means changing the type from "string" to "string | undefined" then let the type reflect reality.',
      },
      {
        selector: 'Literal[value=""]',
        message:
          'Please set the variable to undefined instead. If that means changing the type from "string" to "string | undefined" then let the type reflect reality.',
      },
    ],
```

**Comment by bradzacher** [comment]
I disagree that `''` is equivalent to `string | undefined`. From a pure types perspective - the statement is objectively wrong, but also from a logical perspective `''` is a perfectly valid string, even if it is seen as falsy by implicit boolean coercion.

This rule would provide false safety in your codebase as well, as you it would lead you to assume that "anything that's a `string` cannot be `''`", but that is not the case.

If you decode a JSON blob like `"{\"prop\": \"\"}"`, you would have to ensure `prop` is typed as `string | undefined` (which is nonsensical for a JSON property - which can never be `undefined`).
Similarly if you're dealing with user input, an empty input filed will return `""`, so you have to ensure that is correctly typed as well.

There are many vectors for this rule to lead to complications in your codebase.

Finally - along the lines of incorrect typing, if someone were to just check the nullish case of your string, then you will have broken your assumption and leaked a bug into your code
```ts
const x: string | undefined = '';
if (x != null) {
  getUserById(x); // SAFE, but broken
}
```

---

> I couldn't really figure out how to restrict this rule from firing on only VariableDeclarators that have required string for the typeAnnotation.

```
VariableDeclarator[id.typeAnnotation.typeAnnotation.type = "TSStringKeyword"] > Literal[value=""].init
```

---

Two better solutions for this:

(1) Use runtime invariants. These can be easily compiled out of production code if you want to reduce code weight. These will help you assert in no uncertain terms that input is invalid.

(2) Use branded types.

Branded types allow you to create a special `string` type which a standard `string` cannot be assigned to.
As these are a type-only construct - you can now pass them around without any runtime burden whilst also providing type-check guarantees for your codebase.

```ts
type NonEmptyString = string & { ___brand: unknown };

function assertNotEmptyString(arg: string): asserts arg is NonEmptyString {
  if (arg === '') {
    throw new Error('invariant');
  }
}

function getUserById(id: NonEmptyString){
  return http.get(`/user/${id}`);
}

function main(input: string) {
  getUserById(input); // type error because string is not assignable to NonEmptyString

  assertNotEmptyString(input);
  getUserById(input); // no type error as it's known to be non-empty
}
```
[repl](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gCwxgDgFzxQoCeA3AFBUZn4IByqAogLb50DKGMAligDm8ALzxM-IfABk8AN7wA+soBGMUsGIBXFAGsUSAO4p4AX2pUAZjrAY+qEmjQgYGZhnacyPSYIAUsILEEgKCAJTEUE4uGGgkMMJ8ccwonty8ofJU8PB8lvABCaIiYgDkpWFZOTkY2DBG8CgghvAsMPUwfqUCAG6wfKQYFdQ5plRjVjZ2DoIgGACqzjAAQmQAksB+fJrwKWneGUJhctnwcBhaMCa4BAB0sxh+AAYA9FpLLwAkctumT2HUCbWFC2ewmNhQARbFD4LQYYKHcJVeAPRYuVYbaGwjAA+AvF7wOgMeAuDrwFTgKDvBAhKRJRpILBRNB8QQoKAqCAIDBIXasDjpXw0HLMmLufY+UJYuEA06opYYzYCbG4-EMwn0BCkpAwRy5IZxfRGEw88kIAwoAC0IAFZHGVCAA)

Branded types are also a great way to validate that strings match certain formats (like asserting they're a date, or a phone number, or a UserID, etc)
For example:

```ts
type ValidDateString = string & { ___validDateStringBrand: unknown };

function isValidDateString(str: string): str is ValidDateString {
  return /\d{2}\/\d{2}\/\d{4}/.test(str);
}

function main(input: string) {
  if (isValidDateString(input)) {
    // input is now typed as ValidDateString
  }
}
```

**Comment by dgreene1** [comment]
Thank you for such a well thought out and generous response @bradzacher.

Both of your solutions are great for the purpose of marrying the runtime and compile time types.

What I’m looking for is make it so I don’t have to read code reviews quite as heavily. I am leading multiple teams and the rule I’m proposing forces the development team to think harder about control flow.

But I agree with everything you said. I’m just looking for a rule that will put the onus on my contractor teams to really be certain that about their intentions and to exhaustively cover their cases.

**Comment by bradzacher** [comment]
ESLint rules are easy to write and it's easy to create rules that are local and private to your project.
If you've got some standards that you'd like to automatically enforce for your codebase - I'd suggest creating a local rule.
https://eslint.org/docs/developer-guide/

However - as it stands I don't think that this suggestion makes sense in a general-purpose plugin like this one.

**Comment by dgreene1** [comment]
On further reflection, I agree my proposed rule isn’t a fantastic candidate for the general purpose typescript eslint codebase. I wanted to leave it open to see if other people are annoyed by devs using this kind of “initialize to empty” approach. But again, after thinking about my request (and your excellent response) I would likely have self-closed the issue since my proposal is too subjective of a rule and should be a local project rule only. Thanks again for the conversation and consideration.
